File: 백준/Silver/14940. 쉬운 최단거리/쉬운 최단거리.js
const fs = require("fs");

const input =
  process.platform === "linux"
    ? fs.readFileSync("/dev/stdin").toString().split("\n")  
    : fs.readFileSync("example.txt").toString().split("\r\n");


const dx = [-1,0,1,0]
const dy = [0,1,0,-1]

const [num,...a] = input
const [n,m] = num.split(" ").map((e)=> parseInt(e))
const arr = a.map((v) => v.split(" ").map((e) => parseInt(e)))
const visited = Array.from(Array(n), () => Array(m).fill(0))

const q = []

for (let i=0; i<n; i++) {
    for (let j=0; j < m; j++) {
        if (arr[i][j] == 2) {
            q.push([i,j])
            visited[i][j] = 1
        }
    }
}

while (q.length > 0) {
    const [x,y] = q.shift()

    for (let i=0; i < 4; i++) {
        const nx = x + dx[i]
        const ny = y + dy[i]

        if (0 <= nx && nx < n && 0 <= ny && ny < m) {
            if (arr[nx][ny] == 1 && visited[nx][ny] == 0) {
                q.push([nx,ny])
                visited[nx][ny] = visited[x][y] + 1
            }
        }
    }
}

for (let i=0; i<n; i++) {
    let result = []
    for (let j=0; j < m; j++) {
        if (visited[i][j]) {
            result.push(visited[i][j]-1)
        } else {
            if (arr[i][j]) {
                result.push(-1)
            } else {
                result.push(0)
            }
        }
    }
    console.log(result.join(" "))
}

File: 백준/Silver/14940. 쉬운 최단거리/쉬운 최단거리.py
import sys

input = sys.stdin.readline

from collections import deque

dx = [-1,0,1,0]
dy = [0,1,0,-1]

n,m = map(int,input().split())

arr = [list(map(int,input().split())) for _ in range(n)]


visited = [[0]*m for _ in range(n)]

q = deque()

for i in range(n):
    for j in range(m):
        if arr[i][j] == 2:
            q.append((i,j))
            visited[i][j] = 1

while q:
    x,y = q.popleft()

    for i in range(4):
        nx = x + dx[i]
        ny = y + dy[i]

        if 0 <= nx < n and 0 <= ny < m:
            if arr[nx][ny] == 1 and visited[nx][ny] == 0:
                q.append((nx,ny))
                visited[nx][ny] = visited[x][y] + 1

for i in range(n):
    for j in range(m):
        if visited[i][j]:
            print(visited[i][j] -1 , end=' ')
        else:
            if arr[i][j] == 1:
                print(-1, end=' ')
            else:
                print(visited[i][j], end=' ')
    print()